src/main/java/com/shapesecurity/shift/es2016/codegen/location/LocationMeta.java

```diff
@@ -0,0 +1,68 @@
+package com.shapesecurity.shift.es2016.codegen.location;
+
+import com.shapesecurity.shift.es2016.ast.ImportDeclarationExportDeclarationStatement;
+import com.shapesecurity.shift.es2016.ast.LiteralNumericExpression;
+import com.shapesecurity.shift.es2016.ast.Module;
+import com.shapesecurity.shift.es2016.ast.Node;
+import com.shapesecurity.shift.es2016.ast.Script;
+import com.shapesecurity.shift.es2016.ast.SwitchCase;
+import com.shapesecurity.shift.es2016.ast.SwitchDefault;
+import com.shapesecurity.shift.es2016.codegen.TokenStream;
+import com.shapesecurity.shift.es2016.parser.SourceLocation;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.IdentityHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class LocationMeta {
+	protected final Map<Node, SourceLocation> nodeToStart = new IdentityHashMap<>();
+	protected final Map<Node, SourceLocation> nodeToFinish = new IdentityHashMap<>();
+	protected final List<Node> finishingStatements = new LinkedList<>();
+	protected final List<Node> startingNodes = new LinkedList<>();
+
+	@Nullable
+	protected LiteralNumericExpression lastNumberNode;
+
+	public void startEmit(@NotNull Node node, @NotNull TokenStream ts) {
+		if (!(ts instanceof TokenStreamWithLocation)) return;
+		if (node instanceof Script || node instanceof Module) {
+			this.nodeToStart.put(node, ((TokenStreamWithLocation) ts).getLocation());
+		} else {
+			this.startingNodes.add(node);
+		}
+	}
+
+	public void finishEmit(@NotNull Node node, @NotNull TokenStream ts) {
+		if (!(ts instanceof TokenStreamWithLocation)) return;
+		this.nodeToFinish.put(node, ((TokenStreamWithLocation) ts).getLocation());
+		if (node instanceof ImportDeclarationExportDeclarationStatement || node instanceof SwitchCase || node instanceof SwitchDefault) {
```

Comment:
These are precisely the nodes which might end with an optional semicolon which wouldn't be inserted until the next node codegens. (Although looking at it again I guess I missed `Directive`.)

The case clauses are there because in
```js
switch(0){case 1:foo;case 2:bar}
```

the semicolon should be included in the location of the first case clause.